src/BotPlutusInterface/CoinSelection.hs

```diff
@@ -0,0 +1,497 @@
+    unless (isSufficient outVec finalTxInputVector) $ throwError "Insufficient Funds"
+
+    selectedTxIns <- hoistEither $ mapM txOutToTxIn selectedUtxos
+
+    lift $ printBpiLog @w (Debug [CoinSelectionLog]) $ "Selected TxIns: " <+> pretty selectedTxIns
+
+    -- Now we add the selected utxos to originalTxIns present in the transaction previously.
+    pure $ originalTxIns <> Set.fromList selectedTxIns
+  where
+    -- This represents the condition when we can stop searching for utxos.
+    -- First condition is that the input vector must not be zero vector, i.e.
+    -- There must be at least some input to the transaction.
+    -- Second condition is that all the values of input vector must be greater than
+    -- or equal to the output vector.
+    isSufficient :: ValueVector -> ValueVector -> Bool
+    isSufficient outVec txInsVec =
+      Vec.all (== True) (Vec.zipWith (<=) outVec txInsVec)
+        && txInsVec /= zeroVec (length txInsVec)
+
+-- `searchTxIns` searches for optimal utxos for a transaction as input given
+-- current input vector, output vector and a list of all the remaining utxo vectors.
+searchTxIns ::
+  forall (w :: Type) (effs :: [Type -> Type]).
+  Member (PABEffect w) effs =>
+  SearchStrategy -> -- search strategy to use for selecting utxos
+  (ValueVector -> Bool) -> -- condition on when to stop the search
+  ValueVector -> -- output value vector of the Tx.
+  ValueVector -> -- input value vector of the Tx.
+  [ValueVector] -> -- all the value vectors of the utxos that can be spent.
+  Eff effs (Either Text [Int])
+searchTxIns Greedy stopSearch outVec txInsVec utxosVec =
+  printBpiLog @w (Debug [CoinSelectionLog]) "Selecting UTxOs via greedy search"
+    >> greedySearch @w stopSearch outVec txInsVec utxosVec
+searchTxIns GreedyApprox stopSearch outVec txInsVec utxosVec =
+  printBpiLog @w (Debug [CoinSelectionLog]) "Selecting UTxOs via greedy approx search"
+    >> greedyApprox @w stopSearch outVec txInsVec utxosVec
+
+-- `greedySearch` searches for utxos vectors for input to a transaction,
+-- this is achieved by selecting the utxo vector that have closest euclidean distance
+-- from output vector.
+greedySearch ::
+  forall (w :: Type) (effs :: [Type -> Type]).
+  Member (PABEffect w) effs =>
+  (ValueVector -> Bool) -> -- condition on when to stop the search
+  ValueVector -> -- output value vector of the Tx.
+  ValueVector -> -- input value vector of the Tx.
+  [ValueVector] -> -- all the value vectors of the utxos that can be spent.
+  Eff effs (Either Text [Int])
+greedySearch stopSearch outVec txInsVec utxosVec
+  -- we stop the search if there are no utxos vectors left, as we will not be able to
+  -- select any further utxos as input to a transaction.
+  | null utxosVec =
+    printBpiLog @w (Debug [CoinSelectionLog]) "Greedy: The list of remaining UTxO vectors in null."
+      >> pure (Right mempty)
+  -- we stop the search is the predicate `stopSearch` is true.
```

Comment:
Nit: s/is/if (also applies to similar comment in `greedyapprox`

Reply:
It's a typo, Thanks for pointing it out.